### ISSUE TITLE
How to properly utilize Flair Embeddings within Keras for CNN/LSTM classification? 

### ISSUE THREAD

**Issue by Hellisotherpeople** [flairNLP/flair#987]
So, I've written a lot of code which utilizes Flair to create my word/document vectors - but with sklearn or Keras to do classification with these word-vectors as input. 

My main motivation for this is that my projects involve sklearn pipelines, and I found skorch unsuitable for the task. 

Let's say that my input data is shaped (137, 400) where it's 137 examples each with a 400 dimensional word vector.  The following simple MLP model will work on this input data. 

```python
def create_model(optimizer='adam', kernel_initializer='glorot_uniform', epochs = 5):
        model = Sequential()
        model.add(Dense(list_of_embeddings[1].size, activation='relu',kernel_initializer='he_uniform', use_bias = False))
        model.add(Dense(len(np.unique(Y_val)),activation='softmax',kernel_initializer=kernel_initializer, use_bias = False))
        model.compile(loss='categorical_crossentropy',optimizer=optimizer, metrics=['accuracy'])
        return model
```

But my question is this: **How do I make my input data work with Keras LSTMs or 1D-CNNs?**
Note - I want to **avoid** creating any separate keras embeddings (as is common in CNN tutorials utilizing Keras). Keras's embedding class seems like it will not play nicely with Flair. 

What I've tried
```python
def create_model(optimizer='adam', kernel_initializer='glorot_uniform', epochs = 5):
        model = Sequential()
        model.add(Reshape((137, 1, 400), input_shape = (137, 400))) ##DOESNT WORK! :( 
        #model.add(Conv1D(64, 1, activation='relu'))  #fails, yes there should be pooling after but doesn't even get that far 
        model.add(Dense(list_of_embeddings[1].size, activation='relu',kernel_initializer='he_uniform', use_bias = False))
        model.add(LSTM(list_of_embeddings[1].size, return_sequences = True,)) ##NOT WORKING
        model.add(Dense(len(np.unique(Y_val)),activation='softmax',kernel_initializer=kernel_initializer, use_bias = False))
        model.compile(loss='categorical_crossentropy',optimizer=optimizer, metrics=['accuracy'])
        return model
```

Any ideas? I think that my issues likely have a simple fix but I've been pulling my hair out with the reshape functions not doing what I expect them to do 

The context is provided in my repo where I am trying to enable LSTM/CNN models: 
https://github.com/Hellisotherpeople/Active-Explainable-Classification

**Comment by alanakbik** [comment]
Thanks for posting this. I'd be interested to hear how and how well this works, so if anybody has a solution, please share. 

@Hellisotherpeople how well does the classification with SKlearn work for you?

**Comment by Hellisotherpeople** [comment]
I actually got this working the other day. 

For anyone wondering - Keras does some stupid BS with creating empty "Batch" dimensions...
The solution for getting it to take Flair vectors as input took me so long to figure out and is only intuitive to folks who are much better with Keras than I am. 

Assume input is in the form (153, 300) (153 examples, each a 300 dim vector) 

```python
model.add(Reshape((1, list_of_embeddings[1].size), input_shape = Emb_train.shape[1:])) #Emb_train.shape[1:] is (300,)
```

@alanakbik 

SKLearn classification works quite well. I actually am interested in sklearn compatibility due to my works emphasis on machine learning interpretability (LIME) and class imbalanced datasets. I needed to have flair play nice with sklearns pipelines - as those pipelines are what enable the LIME algorithm to work. To my knowledge, no one has applied model interpretability techniques to modern word embedding powered models. Maybe these a conference paper there somewhere? 

I have found success in the field with sklearn SVM's, as they support class weighting schemes (making them good for imbalanced datasets) and online/incremental learning.

**Comment by alanakbik** [comment]
@Hellisotherpeople cool thanks for sharing the solution - the interpretability works sounds interesting there's surely a paper there :)

**Comment by IsaacCorley** [comment]
@Hellisotherpeople Only somewhat related, but you mentioned using SVMs because of their class weighting parameter; I'm not sure if you are aware, but you can actually set class weights when training Keras models like so

`
class_weight = {0: 1, 1: 50}
model.fit(X_train, Y_train, class_weight=class_weight)
`

You can also compute the balanced class weight using sklearn's compute_class_weight function.

I'm in a similar boat at my current position focusing on black box model interpretability with LIME and imbalanced dataset. Hope this helps.

**Comment by Hellisotherpeople** [comment]
@IsaacCorley 
Thank you for informing me of this! I actually had no ideas that it supported class weighting and will really enjoy experimenting with this on my own projects. 

For you and @alanakbik, my next question is this: 

How would I best incorporate these embeddings into my Keras models? I've already written code to utilize DocumentPoolEmbeddings as the input. I'm much more curious about alternative input strategies... maybe giving it the individual embedding for each word? What about trying to replicate to DocumentLSTMEmbeddings utilizing the Keras embedding layer? I'm sort of worried that a 1D CNN applied on top of already avg'd embeddings may not really be that useful given what a CNN is supposed to do. I've observed that currently LSTM and CNN models do not currently improved performance over that of a MLP when the input is pooled vectors.

**Comment by alanakbik** [comment]
@Hellisotherpeople yes using the word embeddings themselves directly in Keras for classification sounds like a good idea. I haven't experimented with this yet so I don't have any code snippets, but would be very interested to hear how well this works and if it outperforms out own classification layers.

**Comment by Hellisotherpeople** [comment]
```python
'''
with open('card_classification.csv') as csvfile:
    reader = csv.reader(csvfile)
    list_of_sentences = []
    list_of_labels = []
    list_of_embeddings = []
    for row in reader:
        if multi_label:
            list_of_labels.append(row[:-1])
            parsed_string = row[-1]
        else:
            list_of_labels.append(row[0])
            parsed_string = row[1]
        list_of_sentences.append(parsed_string)
        #set_obj = Sentence(parsed_string)
        #stacked_embeddings.embed(set_obj)
        #list_of_embeddings.append(set_obj.get_embedding().cpu().detach().numpy())


# create the tokenizer
t = Tokenizer()
# fit the tokenizer on the documents
t.fit_on_texts(list_of_sentences)

new_labels = b_encoder.fit_transform(list_of_labels)

X_train, X_val, Y_train, Y_val = train_test_split(np.asarray(list_of_sentences), new_labels, test_size = 0.2, stratify = list_of_labels, random_state=42)    
output_size = len(set(list_of_labels))
print(output_size)
print(Y_val[1])
sequences_train = t.texts_to_sequences(X_train)
sequences_valid = t.texts_to_sequences(X_val)



X_train = pad_sequences(sequences_train)
X_val = pad_sequences(sequences_valid, maxlen=X_train.shape[1])
sequence_length = X_train.shape[1]

print(X_train)

# summarize what was learned
#print(encoded_docs)
encoded_docs = t.texts_to_sequences(list_of_sentences)
print("dictionary size: ", len(t.word_index))

vocabulary_size = len(t.word_index)
EMBEDDING_DIM=300


vocabulary_size=len(t.word_index)+1
embedding_matrix = np.zeros((vocabulary_size, EMBEDDING_DIM))


for word, i in t.word_index.items():
    try:
        word_sent = Sentence(word)
        en_embedding.embed(word_sent)
        embedding_vector = word_sent[0].embedding.cpu().detach().numpy()
        embedding_matrix[i] = embedding_vector
    except KeyError:
        embedding_matrix[i]=np.random.normal(0,np.sqrt(0.25),EMBEDDING_DIM)





from keras.layers import Embedding
embedding_layer = Embedding(vocabulary_size,
                            EMBEDDING_DIM,
                            weights=[embedding_matrix],
                            trainable=False)



model = Sequential()
model.add(embedding_layer)
model.add(LSTM(300, return_sequences=False
           , input_shape=(sequence_length, EMBEDDING_DIM)))
model.add(Dense(output_size, activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])
print(model.summary())

model.fit(X_train,Y_train,epochs=20, batch_size=25, verbose=1)
```

For now, proof that keras can utilize Flair embeddings

The tricky part is going to be to get LIME/TextExplainer to be capable of explaining predictions made by this model.

**Comment by BernierCR** [comment]
Woah thank you all. I'm currently happy using the built in classifier, but could easily see myself trying to use Keras in the near future.

**Comment by PradyumnaGupta** [comment]
Use a custom data generator in keras . It will not only help in saving memory while using extremely large size embeddings but also help make any changes/transformations you want with flair embeddings .
